### ISSUE TITLE
Constant lookup in MRuby is diffrent than CRuby

### ISSUE THREAD

**Issue by skandhas** [mruby/mruby#600]
Example：

``` ruby
  class TestForConst
    VERSION = '1.0'

    def self.version1
      VERSION 
    end   

    class << self
      def version2
        VERSION 
      end    
    end

  end  

p TestForConst.version1  
p TestForConst.version2  
```

Ruby 1.9.x output

``` ruby
"1.0"
"1.0"
```

MRuby output

``` ruby
"1.0"
uninitialized constant VERSION (NameError)
```

**Comment by jbreeden** [comment]
This issue is rather old, so just following up with testing. 2/3 cases are passing now (including the original issue).

Original issue seems to be fixed on master

Test:
```Ruby
>   class TestForConst
*     VERSION = '1.0'
*
*     def self.version1
*       VERSION
*     end
*
*     class << self
*       def version2
*         VERSION
*       end
*     end
*
*   end
 => :version2
>
 => nil
> p TestForConst.version1
"1.0"
 => "1.0"
> p TestForConst.version2
"1.0"
 => "1.0"
```

@masamitsu-murase's comment still seems to be valid (mruby const lookup appears based on actual module nesting, while CRuby depends on lexical nesting)

```Ruby
> CONST = "GLOBAL"
 => "GLOBAL"
>
 => nil
> class A
*   CONST = "A"
* end
 => "A"
>
 => nil
> class A::B
*   p CONST
* end
"A"
 => "A"


C:\projects\mruby\bin>irb
DL is deprecated, please use Fiddle
irb(main):001:0> CONST = "GLOBAL"
=> "GLOBAL"
irb(main):002:0>
irb(main):003:0* class A
irb(main):004:1>   CONST = "A"
irb(main):005:1> end
=> "A"
irb(main):006:0>
irb(main):007:0* class A::B
irb(main):008:1>   p CONST
irb(main):009:1> end
"GLOBAL"
=> "GLOBAL"
```

@2peta's issue also appears fixed

```Ruby
> class A
*   C=1
*
*   def self.run
*     p C
*     (0..0).each {|i|
*       p C
*     }
*   end
* end
 => :run
>
 => nil
> A.run
1
1
 => 0..0
```

**Comment by christopheraue** [comment]
Today, I came across this issue, too. Constant lookup is still based on actual nesting:

```
$ cd mruby
$ git log -1
commit b9f771dc50653451d945154671f529903a329033
Author: Yukihiro "Matz" Matsumoto <matz@ruby-lang.org>
Date:   Wed Jun 7 13:07:11 2017 +0900

    Handles exceptions from code generation phase; fix #3695
$ irb
> CONST = "GLOBAL"
 => "GLOBAL"
> 
 => nil
> class A
*   CONST = "A"
* end
 => "A"
> 
 => nil
> class A::B
*   p CONST
* end
"A"
 => "A"
```

**Comment by christopheraue** [comment]
The remaining issue was reported again in #3200. This can be closed imo.
